### PR TITLE
purescript: unbreak package by fixing imports for new mtl

### PR DESCRIPTION
**Pull request by Profpatsch**
mtl decided to un-export `Control.Monad` and `Data.Monoid`, which in itself is fine, but of course it breaks everything.

ATTN: this is based on 24.05, I have no clue how this stuff gets integrated into stable. staging-next? I haven’t yet tried to upstream it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Profpatsch** [comment]
upstream PR: https://github.com/purescript/purescript/pull/4547

**Review by maralorn — COMMENTED** [review]
Left a few remarks. In general with a lot of wiggle room on stable because we don’t need to deal with the code for long anyways.

**Comment by maralorn** [comment]
Ah, the correct branch to merge this into is release-24.05. We can merge it directly there I don’t expect this change to need staging. (Make sure to convert to draft before redirecting the PR, just to be sure.)

**Comment by Profpatsch** [comment]
Actually this change should go on master as well, cause it’s broken there just the same.

**Comment by maralorn** [comment]
Well, then the usual workflow is to figure that change out on the haskell-updates branch first and then do a backport.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 1]
```diff
@@ -2761,6 +2761,8 @@ self: super: {
         lib.pipe
           (super.purescript.overrideScope purescriptOverlay)
           ([
+            # https://github.com/purescript/purescript/pull/4547
+            (appendPatches [./patches/purescript-import-fix.patch])
```
[the remark this answers is not in the document]
Comment (Profpatsch) [reply pkgs/development/haskell-modules/configuration-common.nix:2765]: When I did these, I hadn’t checked whether they are upstreamable yet.

**Comment by Profpatsch** [comment]
@maralorn this applied without merge conflicts on `haskell-updates`: https://github.com/NixOS/nixpkgs/pull/329200

**Pull request merged**